File: backend_kapda_junction/src/models/ReturnRequest.js
const mongoose = require('mongoose');

const returnLineSchema = new mongoose.Schema({
  orderItemId: { type: mongoose.Schema.Types.ObjectId, default: null },
  itemIndex: { type: Number, default: null, min: 0 },
  quantity: { type: Number, required: true, min: 1 }
}, { _id: false });

const RETURN_STATUSES = [
  'requested',
  'approved',
  'rejected',
  'cancelled',
  'pickup_scheduled',
  'in_transit',
  'received',
  'refund_processing',
  'refunded',
  'exchange_processing',
  'exchange_shipped',
  'completed'
];

const returnRequestSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  order: { type: mongoose.Schema.Types.ObjectId, ref: 'Order', required: true },
  type: { type: String, enum: ['return', 'exchange', 'order_cancel'], required: true },
  /** Empty when type is order_cancel (full order). */
  items: { type: [returnLineSchema], default: [] },
  reason: { type: String, required: true },
  reasonDetail: { type: String, default: '' },
  /** Proof video (Cloudinary URL). Required when store policy mandates it. */
  videoUrl: { type: String, default: '' },
  exchangeFor: {
    size: { type: String, default: '' },
    color: { type: String, default: '' }
  },
  status: { type: String, enum: RETURN_STATUSES, default: 'requested' },
  adminNote: { type: String, default: '' },
  rejectReason: { type: String, default: '' },
  pickupCourier: { type: String, default: '' },
  pickupAwb: { type: String, default: '' },
  pickupScheduledAt: { type: Date, default: null },
  refundAmount: { type: Number, default: null },
  /** When exchange is shipped, optional AWB for replacement */
  exchangeAwb: { type: String, default: '' }
}, { timestamps: true });

const TERMINAL = ['rejected', 'refunded', 'completed', 'cancelled'];

returnRequestSchema.pre('validate', function validateItems(next) {
  if (this.type !== 'order_cancel' && (!this.items || this.items.length === 0)) {
    this.invalidate('items', 'At least one line item is required');
  }
  next();
});

returnRequestSchema.index({ order: 1, user: 1, createdAt: -1 });
returnRequestSchema.index({ status: 1 });

const ReturnRequestModel = mongoose.model('ReturnRequest', returnRequestSchema);
ReturnRequestModel.RETURN_STATUSES = RETURN_STATUSES;
ReturnRequestModel.TERMINAL_RETURN_STATUSES = TERMINAL;
module.exports = ReturnRequestModel;

File: flutter_kapda_junction/lib/domain/entities/return_request.dart
import 'package:equatable/equatable.dart';

class ReturnLine extends Equatable {
  final String? orderItemId;
  final int? itemIndex;
  final int quantity;

  const ReturnLine({
    this.orderItemId,
    this.itemIndex,
    required this.quantity,
  });

  @override
  List<Object?> get props => [orderItemId, itemIndex, quantity];
}

class ReturnRequest extends Equatable {
  final String id;
  final String orderId;
  final String type;
  final String status;
  final String reason;
  final String reasonDetail;
  final String videoUrl;
  final String exchangeSize;
  final String exchangeColor;
  final List<ReturnLine> items;
  final String adminNote;
  final String rejectReason;
  final String pickupCourier;
  final String pickupAwb;
  final double? refundAmount;
  final String exchangeAwb;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ReturnRequest({
    required this.id,
    required this.orderId,
    required this.type,
    required this.status,
    required this.reason,
    required this.reasonDetail,
    required this.videoUrl,
    required this.exchangeSize,
    required this.exchangeColor,
    required this.items,
    required this.adminNote,
    required this.rejectReason,
    required this.pickupCourier,
    required this.pickupAwb,
    this.refundAmount,
    required this.exchangeAwb,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isTerminal {
    const t = ['rejected', 'refunded', 'completed', 'cancelled'];
    return t.contains(status);
  }

  @override
  List<Object?> get props => [id];
}
